src/focus-change.cpp

```diff
@@ -155,19 +152,124 @@ class wayfire_focus_change_t : public wf::plugin_interface_t
+                {
+                  case orientation_t::UP:
+                    cond     = align_x;
+                    distance = pos.y - cmp_pos.y;
+                    break;
+
+                  case orientation_t::DOWN:
+                    cond     = align_x;
+                    distance = cmp_pos.y - pos.y;
+                    break;
+
+                  case orientation_t::RIGHT:
+                    cond     = align_y;
+                    distance = cmp_pos.x - pos.x;
+                    break;
+
+                  case orientation_t::LEFT:
+                    cond     = align_y;
+                    distance = pos.x - cmp_pos.x;
+                    break;
+
+                  default: /* unreachable */
+                    ;
+                }
+
+                if (cond && (distance > 0) && (distance < closest_output))
+                {
+                    closest_output = distance;
+                    output = op;
+                }
+            }
+
+            if ((output != nullptr))
             {
-                if (distance < closest_cur)
+                const auto cur_ws = output->wset()->get_current_workspace();
+
+                auto views = output->wset()->get_views(0, std::optional{cur_ws});
+                int32_t closest_cur = INT32_MAX;
+                for (auto&& view : std::move(views))
                 {
-                    closest_cur = distance;
-                    new_focus   = view.get();
+                    const auto bb = view.get()->get_bounding_box();
+                    const auto cx = bb.x + bb.width / 2;
+                    const auto cy = bb.y + bb.height / 2;
+                    if ((cx < 0) || (cx >= pos.width) || (cy < 0) || (cy >= pos.height))
+                    {
+                        continue;
+                    }
+
+                    int32_t distance = INT32_MAX;
+                    switch (orientation)
+                    {
+                      case orientation_t::UP:
+                        distance = pos.height - cy;
+                        break;
+
+                      case orientation_t::DOWN:
+                        distance = cy;
+                        break;
+
+                      case orientation_t::RIGHT:
+                        distance = cx;
+                        break;
+
+                      case orientation_t::LEFT:
+                        distance = pos.width - cx;
+                        break;
+
+                      default: /* unreachable */
+                        ;
+                    }
+
+                    if ((distance >= 0) && (distance < closest_cur))
+                    {
+                        closest_cur = distance;
+                        new_focus   = view.get();
+                    }
                 }
             }
         }
 
         if (new_focus != nullptr)
         {
-            wf::get_core().seat->focus_output(new_focus->get_output());
+            auto op = new_focus->get_output();
```

Comment:
As a 'gotcha', view->get_output(); can return NULL, so we need to handle this case.

Reply:
Nice I will fix that